cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,8 +1273,92 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an `is_in` value set against an inequality guarantee.
+  ///
+  /// Simplifying an `is_in` predicate involves filtering out any values from
+  /// the value set that cannot possibly be found given the guarantee. For
+  /// example, if we have the predicate 'x is_in [1, 2, 3, 4]' and the guarantee
+  /// 'x > 2', then the simplified predicate 'x is_in [3, 4]' is equivalent.
+  /// This can be done efficiently if the value set is sorted and unique by
+  /// binary searching the inequality gound and slicing the value set
+  /// accordingly.
+  ///
+  /// \pre `guarantee` is non-nullable
+  /// \pre `guarantee.bound` is a scalar
+  /// \pre `guarantee.bound.type()->id() == value_set->type_id()`
+  /// \pre `value_set` is non-empty
+  /// \return a simplified value set, or a bool if the simplification results in
+  ///   a boolean literal predicate.
+  template <typename ArrowType>
+  static std::variant<std::shared_ptr<Array>, bool> SimplifyIsInValueSet(
+      const Inequality& guarantee, std::shared_ptr<Array> value_set) {
+    using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+    using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
+    using CType = decltype(checked_pointer_cast<ArrayType>(value_set)->Value(0));
+
+    DCHECK(guarantee.bound.is_scalar());
+    DCHECK_EQ(guarantee.bound.type()->id(), value_set->type_id());
+    DCHECK_GT(value_set->length(), 0);
+
+    CType bound;
+    if constexpr (std::is_same_v<std::shared_ptr<Buffer>,
+                                 typename ScalarType::ValueType>) {
+        bound = static_cast<CType>(*guarantee.bound.scalar_as<ScalarType>().value);
+    } else {
+        bound = guarantee.bound.scalar_as<ScalarType>().value;
+    }
+
+    auto compare = [&bound, &value_set](size_t i) -> Comparison::type {
+      DCHECK(value_set->IsValid(i));
+      auto value = checked_pointer_cast<ArrayType>(value_set)->Value(i);
+      return value == bound ? Comparison::EQUAL
+                            : value < bound ? Comparison::LESS
+                                            : Comparison::GREATER;
+    };
+
+    size_t lo = 0;
+    size_t hi = value_set->length();
+    while (lo + 1 < hi) {
+      size_t mid = (lo + hi) / 2;
+      Comparison::type cmp = compare(mid);
+      if (cmp & Comparison::LESS_EQUAL) {
+        lo = mid;
+      } else {
+        hi = mid;
+      }
```

Comment:
I suspect this binary-search doesn't necessarily work for all possible comparison operators. I don't have time to think carefully about this yet though.